### DISCUSSION TITLE
How do I a Control to e.g. RelativePanel.LeftOf=""  in ViewModel // How do I change alignement in code behind?

### DISCUSSION THREAD

**Discussion by NoizeDaemon** [AvaloniaUI/Avalonia#10330] [Q&A] [answered]
Hi! I've recently started using AvaloniaUI & I am also still pretty new to MVVM in general... I'm stuck on two things atm:

### 1) Is there a way to bind the Name of Controls/bind a reference to a control by Name?

I'm trying to change relative positions in a `RelativePanel` from my ViewModel (I know thats probably not the MVVM way, so there's question 2):

```
<RelativePanel Grid.Row="1">
	<Rectangle Name="R1" Width="30" Height="30" Fill="White"
		RelativePanel.AlignHorizontalCenterWithPanel="True" RelativePanel.AlignVerticalCenterWithPanel="True"/>
	<Rectangle Name="R2" Width="30" Height="30" Fill="Purple"
		RelativePanel.LeftOf="{Binding LeftOf}"/>
	<Rectangle Name="R3" Width="30" Height="30" Fill="Red"
		RelativePanel.RightOf="R1"/>
</RelativePanel>
```

Binding a string `"R1"` to `LeftOf` does not work (it binds the string, but looking at the debugger `RelativePanel.LeftOf` takes the actual control as value, not the name).

So is there a way I could reference a control by it's name from ViewModel (maybe by means of a ValueConverter?).

Alternatively, can you combine multiple Bindings?
There's binding to controls like `$parent[1]` - if I could do sth like `{Binding $parent[{Binding ChildIndex}]}`, that would solve my problem too 😅

Can the relative position even be changed at runtime?

### 2) What would be the correct MVVM way to do this?

Eventually I want to bind an `ItemsControl` (with a `RelativePanel` as `ItemsPanel`) to a list of classes containing amongst others the relative positions to other items.
So I'd start with R1 in the center of the panel, then add R2 to the right of R1, R3 on above R2, etc.

Any help, tips or good ressources to read would be highly appreciated!

### Update:

Trying to do this in code behind:

```
public partial class MainView : UserControl
    {
        private ItemsControl itemsControl;
        private List<Rectangle> visualNumberItems;

        public MainView()
        {
            InitializeComponent();
            visualNumberItems = new List<Rectangle>();
        }

        void NumberItemsChanged(object? sender, NotifyCollectionChangedEventArgs e)
        {
            visualNumberItems = itemsControl
                .GetLogicalDescendants()
                .OfType<ILogical>()
                .OfType<Rectangle>()
                .ToList();

            if (visualNumberItems.Count == 1)
            {
                RelativePanel.SetAlignVerticalCenterWithPanel(visualNumberItems[0], true);
                RelativePanel.SetAlignHorizontalCenterWithPanel(visualNumberItems[0], true);
            }
            else if (visualNumberItems.Count > 1)
            {
                RelativePanel.SetRightOf(visualNumberItems[visualNumberItems.Count - 1], visualNumberItems[visualNumberItems.Count - 2]);
            }
        }

        void NumberItemsControlLoaded(object? sender, RoutedEventArgs e)
        {
            itemsControl = (ItemsControl)sender;
            ((INotifyCollectionChanged)itemsControl.Items).CollectionChanged += NumberItemsChanged;
        }
    }
```

The values for the first Rectangle change, but do not visually update.
The values for the other Rectangles don't even change (see Debugger screenshots).
All the Rectangles are stuck in the upper left corner.

Here's my new XAML:

```
<ItemsControl Name="numberItemsControl" Grid.Row="1"  Items="{Binding NumberItems}" Loaded="NumberItemsControlLoaded">
		<ItemsControl.ItemsPanel>
			<ItemsPanelTemplate>
				<RelativePanel/>
			</ItemsPanelTemplate>
		</ItemsControl.ItemsPanel>
		<ItemsControl.ItemTemplate>
			<DataTemplate>
				<Rectangle Width="50" Height="50" x:Name="{Binding Name}" Fill="White"
							Tag="{Binding Tag}"/>
			</DataTemplate>
		</ItemsControl.ItemTemplate>
</ItemsControl>
```

If I could get that to work, I could potentially change their position based on the Tag I set in my ViewModel...

![image](https://user-images.githubusercontent.com/42150677/218546867-a7285f77-7c3a-4f1b-ad65-53510716c9b9.png)

![image](https://user-images.githubusercontent.com/42150677/218546890-e0366e8b-8c05-44ac-8f61-71f5d8b0c80c.png)

Also is there a way to get only the last added descendant/visual item? I only found options for the first or, like i did here, all of them...

**Comment by rabbitism** [comment, marked as answer]
You need to use converters to convert values in VM to values consumed by the panel.

**Reply by NoizeDaemon** [reply]
I set up multibinding and a barebones multivalueconverter like so:

```
<DataTemplate>
	<Rectangle Width="50" Height="50" Name="{Binding Name}" Fill="White" Tag="{Binding Tag}">
		<RelativePanel.RightOf>
			<MultiBinding Converter="{StaticResource stringToControlConverter}" ConverterParameter="{x:Type Rectangle}">
				<Binding Path="$parent"/>
				<Binding Path="Neighbour"/>
			</MultiBinding>
		</RelativePanel.RightOf>
	</Rectangle>
</DataTemplate>
```

```
public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
{
    return ((IControl)values[0]).FindControl<Rectangle>((string)values[1]);
}
```

(I am using the parent here, but is there a way to pass the root/the entire view like `Path="$root"`?
I can pass a type as parameter, but it's an object in the interface.. how can I use it in sth like `FindControl<T>()`?)

The bindings show up correctly in the debugger:
![image](https://user-images.githubusercontent.com/42150677/218855706-1100ff52-9466-410a-a066-9212b0aaee5e.png)
(Value is of type Rectangle and Name is the Neighbour, same as it would be when i write it directly in XAML)

Same when I use `RelativePanel.SetRight()` from the code behind:
![image](https://user-images.githubusercontent.com/42150677/218856217-25352757-4c61-4ad8-a45e-f040d3c7283f.png)

But still not working -  all the Rectangles are stuck in the upper left corner..

Is there a way to force the UI to redraw?
That would at least tell me if the bindings "work" & there's a problem with the raising of events or sth..

I'm really at my wit's end here 😅

### EDIT:
I can't get the bindings to even show up correctly in the debugger anymore - making me question my sanity.. Maybe they never did & I just took a screenshot of the code behind scenario.. 😭

**Reply by timunie** [reply]
Can you prepare a minimal sample to share with us? I've no idea where to start helping otherwise.

**Reply by NoizeDaemon** [reply]
https://github.com/NoizeDaemon/MinimalSample/tree/main/MinimalSample

It's not super minimal, but more minimal..
At this point I can't even make binding the target string work.. It's not getting passed to the value converter..
Also PropertyChanged does not fire for some reason..
So I'm gonna read up & try to figure it out..

All I want to do is dynamically change the relative position of a control in a RelativePanel by binding some combination of target control (by name/string) and relative position..

So if you have any tips for the meantime, I'd appreciate it.
You can also look through the sample, but it is even further away from what I had made work at one point 😅

**Reply by timunie** [reply]
Interesting sample! I can also not make it work, even if I see the converter does work. I'll investigate further, but I feel like it may be a bug.

**Reply by NoizeDaemon** [reply]
I did a workaround and it's clunky and for now tied to a named preexisting rectangle...
But after days of desperation, I almost cried tears of joy seeing a few colored rectangles jump around on a screen 🤣
Pushed my changes if you're curious :)

I think the problem with the converter was that I set the "unset" value to something other than `AttachedProperty<object>.UnsetValue`. might take another look at it, but I think setting it in the code behind could be cleaner than having multibindings for all directions/relations.

**Reply by NoizeDaemon** [reply]
Finally got it working with bindings too!

![finally](https://user-images.githubusercontent.com/42150677/219249875-2852e698-70ff-4a6e-a6f2-26b803449a4c.gif)

As I suspected, one of the issues was not sending a correct UnsetValue in the Converter (wrongly used `BindingNotification.UnsetValue` at one point). That confuses the panel 😅
Additionally one needs to "Unset" the previous bindings to avoid conflict (leftOf AND rightOf, etc.) and then force the panel to refresh with `InvalidateMeasure().`
For the lack of a better solution I set a tag for the panel, subscribe to changes in view and then change it when I update my other bindings to trigger the refresh.

I was also looking into writing my own AttachedProperties to handle the "Unsetting" outside of my viewmodel/view: https://github.com/NoizeDaemon/MinimalSample/blob/main/MinimalSample/EasyAlignWith.cs
Wanted to change the RelativePanel properties in the setter of my custom property, so I don't have to unset values in the ViewModel, but that does not work for some reason...
Gonna look into that a bit more.

For now I am just happy it works in 2 ways.

All 3 options require quite a bit code though.
Wish there was a built-in single value converter to get controls by their names, to avoid multibinding..
Also an easier way to change RelativePanel properties would be nice.. if the panel could auto-magically use the latest valid update instead of freezing..
Should also take a look at Fody to make variable definitions less bulky 😅

**Reply by timunie** [reply]
I really think it should InvalidateMeasure on it's own. If I were yo, I'd probably use a Behavior instead of a MultiConverter, but that's a different story. 

If we have a look at Canvas, we find this line: https://github.com/AvaloniaUI/Avalonia/blob/5a86bb2f2f065a47c363cd4c2f9e8ef903a2720e/src/Avalonia.Controls/Canvas.cs#L46

I don't see this line for RelavivePanel. Maybe this is the root cause.

**Reply by NoizeDaemon** [reply]
> You need to use converters to convert values in VM to values consumed by the panel.

Marked as answer, since the converter was necessary.
Solution involved a bug fix in #10361 as well though.

Thank you all for your help :)

If someone else struggles with writing the converter, here's my version:

```
    public class StringToControlConverter : IMultiValueConverter
    {
        public static readonly StringToControlConverter Instance = new();

        public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
        {
            if (values[0] == AvaloniaProperty.UnsetValue || values[1] == AvaloniaProperty.UnsetValue || values[1] is null)
                return BindingOperations.DoNothing;
            else if ((string)values[1] == "")
                return AvaloniaProperty.UnsetValue;
            else
            {
                Control parent = (Control)values[0];
                string target = (string)values[1];

                Control control = parent.FindControl<Control>(target);

                return control ?? AvaloniaProperty.UnsetValue;
            }
        }
    }
```

And here's how I implemented it in my view:

```
<UserControl.Resources>
       <local:StringToControlConverter x:Key="stringToControlConverter"/>
</UserControl.Resources>
```

```
...
<RelativePanel.AlignHorizontalCenterWith>
	<MultiBinding Converter="{StaticResource stringToControlConverter}">
		<Binding Path="$parent"/>
		<Binding Path="AlignHorizontalCenterWith"/>
	</MultiBinding>
</RelativePanel.AlignHorizontalCenterWith>
...
```

Eventually it would be nice to rewrite it to not require multibinding...
